frontend/src/app/id-manager.service.ts

```diff
@@ -0,0 +1,71 @@
+/*Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+import {Injectable} from '@angular/core';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class IdManagerService {
+  nextID = 0;
+  constructor() {}
+
+  /**
+   * Given some sample, assignIDs will return that sample with an ID
+   * prepended to each individual trace.
+   * @param sample The sample that needs IDs assigned to it.
+   */
+  public assignIDs(sample) {
+    const ids = this.getIDs(this.countTraces(sample));
+    sample.timestamp_diffs = [ids.pop(), sample.timestamp_diffs];
+    if ('latencies' in sample) {
+      sample.latencies = [ids.pop(), sample.timestamp_diffs];
+    }
+    for (const i in sample.data) {
+      sample.data[i] = [ids.pop(), sample.data[i]];
```

Comment:
Instead of changing data from [data] to [id, [data]] on the fly, can we set the id with a dummy value (or real id) initially? Thsi way the sample object doesn't change from how it's defined.